cbauthx/cbauth.go

```diff
@@ -0,0 +1,437 @@
+
+	// HeartbeatTimeout specifies the period of time we are willing to
+	// wait between heartbeats before marking the connection dead.
+	HeartbeatTimeout time.Duration
+
+	// LivenessTimeout specifies how long we are willing to keep using the cache
+	// after we are no longer sure that the cache is still valid (how long since
+	// the last revrpc message was received from ns_server).
+	LivenessTimeout time.Duration
+
+	// ConnectTimeout specifies how long to wait for each attempt to connect
+	// cbauth clients.  Clients will be rejected and a new connection attempted
+	// if the connect timeout exceeds this time.
+	ConnectTimeout time.Duration
+
+	// MinReconnectTime specifies the minimum amount of time between attempts
+	// to connect to a node.  If a connection errors and will be retried in less
+	// time than this, cbauth will wait until this time before retrying.
+	MinReconnectTime time.Duration
+}
+
+func NewCbAuth(ctx context.Context, config *CbAuthConfig, opts *CbAuthOptions) (*CbAuth, error) {
+	if opts.LivenessTimeout <= opts.HeartbeatInterval {
+		return nil, errors.New("liveness timeout must be higher than heartbeat interval")
+	}
+
+	connectTimeout := 10 * time.Second
+	if opts.ConnectTimeout > 0 {
+		connectTimeout = opts.ConnectTimeout
+	}
+
+	minReconnectTime := 1 * time.Second
+	if opts.MinReconnectTime > 0 {
+		minReconnectTime = opts.MinReconnectTime
+	}
+
+	auth := &CbAuth{
+		logger:            opts.Logger,
+		httpTransport:     opts.HttpTransport,
+		serviceName:       opts.ServiceName,
+		userAgent:         opts.UserAgent,
+		heartbeatInterval: opts.HeartbeatInterval,
+		heartbeatTimeout:  opts.HeartbeatTimeout,
+		livenessTimeout:   opts.LivenessTimeout,
+		connectTimeout:    connectTimeout,
+		minReconnectTime:  minReconnectTime,
+		endpoints:         config.Endpoints,
+		username:          config.Username,
+		password:          config.Password,
+		clusterUuid:       config.ClusterUuid,
+		reconfigureSigCh:  make(chan struct{}),
+	}
+
+	cli, endpoint, err := auth.newClient(ctx,
+		auth.endpoints[:1],
+		auth.username,
+		auth.password,
+		auth.clusterUuid)
+	if err != nil {
+		return nil, &contextualError{
+			Message: "failed to connect cbauth client",
+			Cause:   err,
+		}
+	}
+
+	auth.currentClient.Store(cli)
+
+	go auth.runThread(
+		cli,
+		endpoint,
+		auth.username,
+		auth.password,
+		auth.clusterUuid)
+
+	return auth, nil
+}
+
+// newClient creates a new client to attempt to use.  Note that this function
+// is not "thread-safe" and must be invoked singularily at a time
+func (a *CbAuth) newClient(
+	ctx context.Context,
+	endpoints []string,
+	username string,
+	password string,
+	clusterUuid string,
+) (*CbAuthClient, string, error) {
+	// Do not attempt to connect to cbauth faster than allowed
+	{
+		lastConnectDelta := time.Since(a.lastConnectTime)
+		if lastConnectDelta < a.minReconnectTime {
+			select {
+			case <-time.After(a.minReconnectTime - lastConnectDelta):
+			case <-ctx.Done():
```

Comment:
Should this do something other than just keep running?

Reply:
Yes.  Good point.